pkg/apis/volumesnapshot/v1beta1/types.go

```diff
@@ -0,0 +1,372 @@
+
+	// source specifies from where a snapshot will be created.
+	// This field is immutable after creation.
+	// Required.
+	Source VolumeSnapshotContentSource `json:"source" protobuf:"bytes,5,opt,name=source"`
+}
+
+// VolumeSnapshotContentSource represents the CSI source of a snapshot.
+// Exactly one of its members must be set.
+// Members in VolumeSnapshotContentSource are immutable.
+// TODO(xiangqian): Add a webhook to ensure that VolumeSnapshotContentSource members
+// will be immutable once specified.
+type VolumeSnapshotContentSource struct {
+	// volumeHandle specifies the CSI name of the volume from which a snapshot
+	// should be dynamically taken from.
+	// This field is immutable once specified.
+	// +optional
+	VolumeHandle *string `json:"volumeHandle,omitempty" protobuf:"bytes,1,opt,name=volumeHandle"`
+
+	// snapshotHandle specifies the CSI name of a pre-existing snapshot on the
+	// underlying storage system.
+	// This field is immutable once specified.
+	// +optional
+	SnapshotHandle *string `json:"snapshotHandle,omitempty" protobuf:"bytes,2,opt,name=snapshotHandle"`
+}
+
+// VolumeSnapshotContentStatus is the status of a VolumeSnapshotContent object
+type VolumeSnapshotContentStatus struct {
+	// snapshotHandle is the CSI name of a snapshot on the underlying storage system.
+	// If not specified, it indicates that dynamic snapshot creation has either failed
+	// or it is still in progress.
+	// +optional
+	SnapshotHandle *string `json:"snapshotHandle,omitempty" protobuf:"bytes,1,opt,name=snapshotHandle"`
+
+	// creationTime is the timestamp when the point-in-time snapshot is taken
+	// by the underlying storage system. This timestamp is returned by the CSI
+	// driver after the snapshot is cut.
+	// The format of this field is a Unix nanoseconds time encoded as an int64.
+	// On Unix, the command `date +%s%N` returns the current time in nanoseconds
+	// since 1970-01-01 00:00:00 UTC.
+	// +optional
+	CreationTime *int64 `json:"creationTime,omitempty" protobuf:"varint,2,opt,name=creationTime"`
+
+	// restoreSize represents the complete size of the snapshot in bytes.
+	// When restoring a volume from this snapshot, the size of the volume MUST NOT
+	// be smaller than the restoreSize if it is specified.
+	// Otherwise the restoration will fail.
+	// If not specified, it indicates that the size is unknown.
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	RestoreSize *int64 `json:"restoreSize,omitempty" protobuf:"bytes,3,opt,name=restoreSize"`
+
+	// readyToUse indicates if a snapshot is ready to be used to restore a volume.
+	// In dynamic snapshot creation case, this field will be filled in with the
+	// value returned from CSI "CreateSnapshotRequest" gRPC call.
+	// For pre-existing snapshot, this field will be updated with the value returned
+	// from CSI "ListSnapshots" gRPC call if the corresponding driver supports.
```

Comment:
Also same comment about being consistent wrt mentioning that we get information from ListSnapshots if the driver supports it